mono/metadata/cominterop.c

```diff
@@ -3699,32 +3669,46 @@ mono_cominterop_get_com_interface (MonoObject *object, MonoClass *ic, MonoError
 			return NULL;
 		klass = mono_object_class (object);
 		if (!mono_class_is_transparent_proxy (klass)) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "Class is not transparent");
 			return NULL;
 		}
 
 		real_proxy = ((MonoTransparentProxy*)object)->rp;
 		if (!real_proxy) {
+			if (icall)
+				g_assert_not_reached ();
```

Comment:
Same.  Put `g_assert (real_proxy)` in `..._internal` and the dynamic check in `mono_cominterop_get_com_interface`